core/src/main/scala/org/apache/spark/executor/ProcfsMetricsSource.scala

```diff
@@ -19,14 +19,15 @@ package org.apache.spark.executor
 
 import com.codahale.metrics.{Gauge, MetricRegistry}
 
+import org.apache.spark.SparkEnv
 import org.apache.spark.internal.config
 import org.apache.spark.metrics.source.Source
-import org.apache.spark.SparkEnv
 
 private[executor] class ProcfsMetricsSource extends Source {
   override val sourceName = "procfs"
-  override val metricRegistry = new MetricRegistry()
+  // We use numMetrics for tracking to only call computAllMetrics once per set of metrics
```

Comment:
Can we remove this now that we have caching? This looks like a hacky way to achieve this anyway, it will be better to have some alternate way.

Reply:
I thought that this still can save us from unnecessary calls. So I kept it. Why you think it is hacky? The way that Metrics system is designed is that it just return a single value from a guage. There are some other methods to return a set of metrics, but to use that we need to make more changes to the procfsgetter to implement Dropwised metric interface for each metric that we are going to report.  I think that isn't necessary and it make the code uglier.

Reply:
The reason I think this is hacky is that `getProcfsMetrics` relies on an internal state (`numMetrics`) to determine whether to call `computeAllMetrics` or not. I will prefer that `getProcfsMetrics` is a stateless method though.

I understand the limitations imposed by the MetricsSystem but would still like you to evaluate alternate approaches. If you think this is the best way to achieve this, then it is fine.

Reply:
I'm also with Ankur here, I don't understand the point of this.  other metrics, eg. `NettyMemoryMetrics` dont' seem to need to do the same thing w/ `numMetrics`, and nothing about the metrics api makes it look like you'd need to.  Why do you think this is necessary?  It seems you just need to return a `MetricRegistry` with all of the metrics registered there.

Reply:
I think this is good to be done here to avoid calling procfsMetricsGetter.computeAllMetrics to compute the same set of metrics

Reply:
BTW, NettyMemoryMetrics is implementing MetricSet and each metrics there also  have implemented Metric interface. As I responded in my earlier comment, If I go that route I will avoid this here, but then code in ProcfsMetricGetter will be much uglier and to be honest I don't want to change that since it took 5 months for us to reach an agreement there. The gain also wouldn't be that much. The purpose of this code here is to have a less impact on the performance by removing unnecessary calls.